File: typescript/104.ts
import { buildTree, TreeNode } from './tree';

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxDepth(root: TreeNode | null): number {
    return root ? Math.max(maxDepth(root.left), maxDepth(root.right)) + 1 : 0;
};

test('104', () => {
    expect(maxDepth(buildTree([3, 9, 20, null, null, 15, 7]))).toBe(3);
    expect(maxDepth(buildTree([1, null, 2]))).toBe(2);
});

File: python/104.py
from typing import Optional
import unittest

from tree import TreeNode, buildTree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))


class TestSolution(unittest.TestCase):
    def test_maxDepth(self):
        solution = Solution()
        self.assertEqual(solution.maxDepth(
            buildTree([3, 9, 20, None, None, 15, 7])), 3)
